### PR TITLE
Support CPP in the IDE

### PR DESCRIPTION
**Pull request by neil-da**
This code _isn't_ to enable CPP in DAML, but to make the IDE support it, so that upstream Haskell programs using the IDE can take advantage of it.

### CODE REVIEW

File: compiler/haskell-ide-core/src/Development/IDE/Functions/CPP.hs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  ,"  (major1) <  ",major1," || \\\n"
+  ,"  (major1) == ",major1," && (major2) <  ",major2," || \\\n"
+  ,"  (major1) == ",major1," && (major2) == ",major2," && (minor) <= ",minor,")"
+  ,"\n\n"
+  ]
+  where
+    (major1:major2:minor:_) = map show (versionBranch version ++ repeat 0)
+
+-- ---------------------------------------------------------------------------
+-- join object files into a single relocatable object file, using ld -r
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Functions/CPP.hs]: Do we want to have this comment here? It looks like you only copied the comment but not the logic the comment refers to.

Comment (neil-da) [reply compiler/haskell-ide-core/src/Development/IDE/Functions/CPP.hs]: Deleted. Thanks

**Comment by neil-da** [comment]
@cocreature this is the last patch in the series that makes it reusable. The next patch is a `.cabal` file and demo, at which point we'll be testing it.

**Comment by cocreature** [comment]
Nice, looking forward to that!

**Comment by hurryabit** [comment]
To clarify, my use case is in `daml-stdlib`. So no need, to expose this as a language feature to our users.

**Comment by neil-da** [comment]
The CPP supplied by GHC calls out to the system C preprocessor and requires setting a lot of flags in the GHC session. The former is probably no big deal in our system libraries. The latter probably requires a different implementation of the GHC Session - but as it happens, other changes probably require a different implementation of the GHC Session (transitive reexported dependencies), so I'll see what I can do.

**Pull request merged**